From 39de4f7d44116fcf821fb14506a9103d02d4017b Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Thu, 10 Oct 2024 16:28:38 +0200
Subject: [PATCH 1317/1695] clk: stm32mp21: handle RIF semaphores when non
 TDCID

Rework the handling of RIF semaphores to be able to separate the CID
filtering configuration from the acquisition of the RIF semaphores
that is now done in acquire_available_semaphores(). When
CFG_STM32_CM33TDCID is set, acquire the semaphore when probing the
driver and before accessing RCC resources.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 333c0e8d4764df84bb97a33ede05d0ffd0667481 clk: stm32mp21: handle RIF semaphores when non TDCID]
Change-Id: I6815d12ce65fce2ffd8861a25195632390748a2c
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/410512
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/clk/clk-stm32mp21.c | 131 ++++++++++++++++++-------------
 1 file changed, 75 insertions(+), 56 deletions(-)

diff --git a/core/drivers/clk/clk-stm32mp21.c b/core/drivers/clk/clk-stm32mp21.c
index 534075521..c0d46d3d4 100644
--- a/core/drivers/clk/clk-stm32mp21.c
+++ b/core/drivers/clk/clk-stm32mp21.c
@@ -97,8 +97,6 @@
 #define RCC_NB_RIF_RES			U(114)
 #define RCC_NB_CONFS			DIV_ROUND_UP(RCC_NB_RIF_RES, 32)
 
-#define RCC_NB_MAX_CID_SUPPORTED	U(7)
-
 /* Register: RCC_RxCIDCFGR */
 #define RCC_CIDCFGR_CFEN		BIT(0)
 #define RCC_CIDCFGR_SEM_EN		BIT(1)
@@ -4313,42 +4311,77 @@ static struct clk_stm32_priv stm32mp21_clock_data = {
 	.is_critical		= clk_stm32_clock_is_critical,
 };
 
-static TEE_Result apply_rcc_rif_config(bool is_tdcid)
+static TEE_Result handle_available_semaphores(void)
 {
-	TEE_Result res = TEE_ERROR_ACCESS_DENIED;
 	struct stm32_clk_platdata *pdata = &stm32mp21_clock_pdata;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned int index = 0;
 	uint32_t cidcfgr = 0;
 	unsigned int i = 0;
-	unsigned int index = 0;
 
 	for (i = 0; i < RCC_NB_RIF_RES; i++) {
+		vaddr_t reg_offset = pdata->rcc_base + RCC_SEMCR(i);
+
 		index = i / 32;
+
 		if (!(BIT(i % 32) & pdata->conf_data.access_mask[index]))
 			continue;
 
-		/*
-		 * When TDCID, OP-TEE should be the one to set the CID filtering
-		 * configuration. Clearing previous configuration prevents
-		 * undesired events during the only legitimate configuration.
-		 */
-		if (is_tdcid)
-			io_clrbits32(pdata->rcc_base + RCC_CIDCFGR(i),
-				     RCC_CIDCFGR_CONF_MASK);
-
 		cidcfgr = io_read32(pdata->rcc_base + RCC_CIDCFGR(i));
 
-		/* Check if the resource is in semaphore mode */
 		if (!stm32_rif_semaphore_enabled_and_ok(cidcfgr, RIF_CID1))
 			continue;
 
-		/* If not TDCID, we want to acquire semaphores assigned to us */
-		res = stm32_rif_acquire_semaphore(pdata->rcc_base +
-						  RCC_SEMCR(i),
-						  RCC_NB_MAX_CID_SUPPORTED);
-		if (res) {
-			EMSG("Could not acquire semaphore for resource %u", i);
-			return res;
+		if (!(io_read32(pdata->rcc_base + RCC_SECCFGR(index)) &
+		      BIT(i % 32))) {
+			res = stm32_rif_release_semaphore(reg_offset,
+							  MAX_CID_SUPPORTED);
+			if (res) {
+				EMSG("Cannot release semaphore for resource %"PRIu32,
+				     i);
+				return res;
+			}
+		} else {
+			res = stm32_rif_acquire_semaphore(reg_offset,
+							  MAX_CID_SUPPORTED);
+			if (res) {
+				EMSG("Cannot acquire semaphore for resource %"PRIu32,
+				     i);
+				return res;
+			}
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result apply_rcc_rif_config(bool is_tdcid)
+{
+	TEE_Result res = TEE_ERROR_ACCESS_DENIED;
+	struct stm32_clk_platdata *pdata = &stm32mp21_clock_pdata;
+	unsigned int i = 0;
+	unsigned int index = 0;
+
+	if (is_tdcid) {
+		for (i = 0; i < RCC_NB_RIF_RES; i++) {
+			index = i / 32;
+			if (!(BIT(i % 32) &
+			      pdata->conf_data.access_mask[index]))
+				continue;
+
+			/*
+			 * When TDCID, OP-TEE should be the one to set the CID
+			 * filtering configuration. Clearing previous
+			 * configuration prevents undesired events during the
+			 * only legitimate configuration.
+			 */
+			io_clrbits32(pdata->rcc_base + RCC_CIDCFGR(i),
+				     RCC_CIDCFGR_CONF_MASK);
 		}
+	} else {
+		res = handle_available_semaphores();
+		if (res)
+			panic();
 	}
 
 	/* Security and privilege RIF configuration */
@@ -4373,36 +4406,6 @@ static TEE_Result apply_rcc_rif_config(bool is_tdcid)
 		io_clrsetbits32(pdata->rcc_base + RCC_CIDCFGR(i),
 				RCC_CIDCFGR_CONF_MASK,
 				pdata->conf_data.cid_confs[i]);
-
-		cidcfgr = io_read32(pdata->rcc_base + RCC_CIDCFGR(i));
-
-		/*
-		 * Take semaphore if the resource is in semaphore mode
-		 * and secured
-		 */
-		if (!stm32_rif_semaphore_enabled_and_ok(cidcfgr, RIF_CID1) ||
-		    !(io_read32(pdata->rcc_base + RCC_SECCFGR(index)) &
-		      BIT(i % 32))) {
-			res =
-			stm32_rif_release_semaphore(pdata->rcc_base +
-						    RCC_SEMCR(i),
-						    RCC_NB_MAX_CID_SUPPORTED);
-			if (res) {
-				EMSG("Could not release semaphore for res%u",
-				     i);
-				return res;
-			}
-		} else {
-			res =
-			stm32_rif_acquire_semaphore(pdata->rcc_base +
-						    RCC_SEMCR(i),
-						    RCC_NB_MAX_CID_SUPPORTED);
-			if (res) {
-				EMSG("Could not acquire semaphore for res%u",
-				     i);
-				return res;
-			}
-		}
 	}
 
 	for (index = 0; index < RCC_NB_CONFS; index++) {
@@ -4411,6 +4414,9 @@ static TEE_Result apply_rcc_rif_config(bool is_tdcid)
 				pdata->conf_data.lock_conf[index]);
 	}
 
+	res = handle_available_semaphores();
+	if (res)
+		return res;
 end:
 	if (IS_ENABLED(CFG_TEE_CORE_DEBUG)) {
 		for (index = 0; index < RCC_NB_CONFS; index++) {
@@ -4474,13 +4480,20 @@ static TEE_Result stm32_rcc_rif_pm(enum pm_op op, unsigned int pm_hint,
 	if (res)
 		return res;
 
-	if (!PM_HINT_IS_STATE(pm_hint, CONTEXT) || !is_tdcid)
+	if (!PM_HINT_IS_STATE(pm_hint, CONTEXT))
 		return TEE_SUCCESS;
 
-	if (op == PM_OP_RESUME)
-		res = stm32_rcc_rif_pm_resume();
-	else
+	if (op == PM_OP_RESUME) {
+		if (!is_tdcid)
+			res = handle_available_semaphores();
+		else
+			res = stm32_rcc_rif_pm_resume();
+	} else {
+		if (!is_tdcid)
+			return TEE_SUCCESS;
+
 		res = stm32_rcc_rif_pm_suspend();
+	}
 
 	return res;
 }
@@ -4576,6 +4589,12 @@ static TEE_Result stm32mp2_clk_probe(const void *fdt, int node,
 		}
 	}
 
+	if (IS_ENABLED(CFG_STM32_CM33TDCID)) {
+		res = handle_available_semaphores();
+		if (res)
+			return res;
+	}
+
 	rc = clk_stm32_init(priv, stm32_rcc_base());
 	if (rc) {
 		res = TEE_ERROR_GENERIC;
-- 
2.47.2

